### PR TITLE
Do not exclude empty arrays or empty objects in source filtering with Jackson streaming

### LINKED ISSUE #109668: Regression in 8.6.0: Not returning empty arrays when source filtering inside them
### Elasticsearch Version

from 8.6.0 until current

### Installed Plugins

_No response_

### Java Version

openjdk version "19.0.1" 2022-10-18

### OS Version

Linux e3b996519639 6.5.0-35-generic #35~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC

### Problem Description
Exclude filtering causes empty objects from the response to disappear entirely.

This is probably very similar to these.
https://github.com/elastic/elasticsearch/issues/29622
https://github.com/elastic/elasticsearch/issues/23796

But the thing is that this WAS working fine in 8.3.x, 8.4.x, 8.5.x, ...

8.6.0 broke this feature again.

#### setup data
```
PUT /test
```

```
PUT /test/_mapping
{
  "properties": {
    "myArray": {
      "properties": {
        "myField": {
          "type": "keyword"
        }
      }
    }
  }
}
```

```
PUT /test/_doc/1
{
  "myArray": [{"myField": "value", "other": "value"}]
}
```

```
PUT /test/_doc/2
{
  "myArray": []
}
```

#### search
```
POST /test/_search
{
  "_source": {
    "includes": ["myArray"],
    "excludes": ["myArray.myField"]
  }
}
```

#### result
```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test",
        "_id": "2",
        "_score": 1,
        "_source": {}
      },
      {
        "_index": "test",
        "_id": "1",
        "_score": 1,
        "_source": {
          "myArray": [
            {
              "other": "value"
            }
          ]
        }
      }
    ]
  }
}
```
It should contain "myArray" with empty arrays.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by mccheah**
Fixes #109668.

This fixes the specific case outlined in the above linked Github issue - namely this:

* Document: `{ "myArray": [] }`
* Source filter:  Include `myArray`, exclude `myArray.myField`
* Expected filtered result: `{ "myArray": [] }`
* Actual filtered result: `{}`

This broke when we switched to from map-based source filtering to Jackson-streaming-based source filtering in some code paths - see #91127. Map-based source filtering (using the automata, etc.) correctly included the empty array from the parent. Unfortunately, it's not clear when these two code paths for source filtering diverge in behavior.

Note also the long comment in `FilterPathBasedFilter`. It's not clear what the correct behavior is when we are filtering out all contents of arrays via source filter path exclusion.

### CODE REVIEW

**Comment by rjernst** [comment]
I think the search team needs to weigh in on whether the existing or old behavior is intended.

**Comment by mccheah** [comment]
I had to also update the same case for empty objects.

@elastic/es-search Any chance someone can take a look at this?

**Comment by mccheah** [comment]
@nik9000 - I noticed you gave the initial +1 for https://github.com/elastic/elasticsearch/pull/91127 which is in this space. Would you please opine on this patch and the expected behavior in this case? Thanks!

**Comment by mccheah** [comment]
Hello - just pinging here again, would it be possible for someone to take a look at this? Thanks!

**Comment by mccheah** [comment]
Hello - checking in on this again. Would it be possible to get a review of this soon?

**Comment by carlosdelest** [comment]
@elasticsearchmachine test this please

**Comment by carlosdelest** [comment]
@elasticsearchmachine update branch

**Comment by carlosdelest** [comment]
@elasticmachine update branch

**Comment by carlosdelest** [comment]
@elasticsearchmachine test this please

**Comment by carlosdelest** [comment]
I've checked the versions, and it is indeed reproducible from 8.6.0. Behaviour on 8.5.0 was as described - I used the reproduction steps in https://github.com/elastic/elasticsearch/issues/109668

**Comment by carlosdelest** [comment]
@elasticmachine generate changelog

**Comment by carlosdelest** [comment]
@elasticsearchmachine generate changelog

**Comment by carlosdelest** [comment]
@elasticsearchmachine test this please

**Comment by mccheah** [comment]
@carlosdelest do I need to do something to fix the build? I tried navigating to the buildkite page but it's giving me a "page not found" error.

**Comment by carlosdelest** [comment]
@elasticsearchmachine test this please

**Comment by carlosdelest** [comment]
> @carlosdelest do I need to do something to fix the build? I tried navigating to the buildkite page but it's giving me a "page not found" error.

No worries, Checkstyle was failing and I fixed it. Next time you can check locally with `./gradlew :libs:elasticsearch-x-content:impl:check` before submitting a change 👍

**Comment by carlosdelest** [comment]
@elasticsearchmachine test this please

**Review by carlosdelest — APPROVED** [review]
LGTM 👍

**Comment by carlosdelest** [comment]
@rjernst , from Search I have tested that this behaviour was changed from the original one. This change reverts to the behaviour in versions prior to 8.6 where the regression was introduced.

It LGTM - please feel free to merge if it looks OK from core infra team 👍

**Comment by carlosdelest** [comment]
@mccheah , there are no current plans to issue another 8.14.x release. I've added a backport for 8.15.3 which is the closer we can get to 8.14.x as of now.

**Comment by mccheah** [comment]
@rjernst A quick ping on this - any chance this can be merged and released on the 8.x line? Thanks!

**Comment by mccheah** [comment]
@rjernst @carlosdelest By any chance, can we get this merged and backported soon? Thanks!

File: server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java [hunk 1]
```diff
@@ -112,4 +112,48 @@ public Source filter(SourceFilter sourceFilter) {
...
+    public void testIncludeParentAndExcludeChildEmptyObject() {
+        Source fromMap = Source.fromMap(Map.of("myObject", Map.of()), XContentType.JSON);
+        Source filteredMap = fromMap.filter(new SourceFilter(new String[] { "myObject" }, new String[] { "myObject.myField" }));
+        assertEquals(filteredMap.source(), Map.of("myObject", Map.of()));
+        Source fromBytes = Source.fromBytes(new BytesArray("{\"myObject\": {}}"), XContentType.JSON);
+        Source filteredBytes = fromBytes.filter(new SourceFilter(new String[] { "myObject" }, new String[] { "myObject.myField" }));
+        assertEquals(filteredBytes.source(), Map.of("myObject", Map.of()));
+    }
+
+    public void testIncludeParentAndExcludeChildSubFieldsArrays() {
```
Comment (prdoyle) [inline server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:134]: Note to self: This test passes even without the fix.

File: server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java [hunk 2]
```diff
@@ -112,4 +112,48 @@ public Source filter(SourceFilter sourceFilter) {
...
+        );
+        Source filteredMap = fromMap.filter(new SourceFilter(new String[] { "myArray" }, new String[] { "myArray.myField" }));
+        assertEquals(filteredMap.source(), Map.of("myArray", List.of(Map.of("other", "otherValue"))));
+        Source fromBytes = Source.fromBytes(new BytesArray("""
+            { "myArray": [ { "myField": "myValue", "other": "otherValue" } ] }"""), XContentType.JSON);
+        Source filteredBytes = fromBytes.filter(new SourceFilter(new String[] { "myArray" }, new String[] { "myArray.myField" }));
+        assertEquals(filteredBytes.source(), Map.of("myArray", List.of(Map.of("other", "otherValue"))));
+    }
+
+    public void testIncludeParentAndExcludeChildSubFieldsObjects() {
```
Comment (prdoyle) [inline server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:147]: Note to self: This test also passes without the fix.

**Review by prdoyle — APPROVED** [review]
Seems reasonable. My only concern is that perhaps two of the four tests aren't testing what @mccheah thought they were, because those tests pass even without this fix.

File: server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java (hunk 1 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:134]: I think @mccheah intention was to test map based filtering for this behaviour, which did work before the fix. Can you please confirm @mccheah ?

**Comment by carlosdelest** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/115222)|
|✅|8.16|[](https://github.com/elastic/elasticsearch/pull/115223)|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/115224)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

File: server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java (hunk 1 above)
Comment (mccheah) [reply server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:134]: Not exactly - though this does inform me I didn't write this in the most intuitive way, and if I were to do this again I'd adjust it -

There's two test cases embedded in this. Assuming no patch: The first does serde from Map, which passes. The second does serde with bytes, which fails.

What I would self-recommend having done instead was creating separate test cases for each of map and byte serde.

Comment (mccheah) [reply server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:134]: Oh wait never mind - In the tests that pass without the fix, I was making sure I didn't introduce regression in more "standard" cases via the patch I had written. Sorry, got mixed up in my mind =)

Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/search/lookup/SourceFilterTests.java:134]: That was my assumption as well 👍

**Pull request merged**